File: src/Lobster.Framework/LobsterClasses/LobsterObject/DomainModel/ComponentSet.vb
'LobsterDb
'Copyright(c) 2010, LobsterDb
'Licensed under the AGPL Version 3 license
'http://www.lobsterdb.com/license

Public Class ComponentSet

    Public _componentModel As ClassModelReader.ClassModel
    Public _parent As Composite
    Private _stateArray As StateArray

    Sub New(ByVal componentModel As ClassModelReader.ClassModel, _
            ByVal stateArray As StateArray, ByVal parent As Composite)
        _componentModel = componentModel
        _stateArray = stateArray
    End Sub

    Public Function GetNew() As Component
        'Dim component As New Component()
        Dim stateObject = _stateArray.AddObject()
        Dim component As New Component(_componentModel, stateObject)
        Return component
    End Function

    Public Function GetCollection() As Collection
        Dim collection As New Collection
        For i As Integer = 0 To _stateArray.GetLength - 1
            collection.Add(New Component(_componentModel, _
                                         _stateArray.GetObject(i)))
        Next
        Return collection
    End Function

    Friend Sub Persist(ByVal connection As LobsterConnection, _
                       ByVal compositeKeyName As String, _
                       ByVal compositeKey As String)
        Dim command As New SqlCommand(SqlBuilder.GetInsertSql(_componentModel), _
                                      connection.Connection)
        For Each component As Component In Me.GetCollection()
            component.Persist(command, connection, compositeKeyName, compositeKey)
        Next
    End Sub

End Class

File: src/LobsterDb.Client/Composites/Composite/ComponentSet.js
/*!
 * LobsterDb
 * Copyright(c) 2010, LobsterDb
 * Licenced under the AGPL Version 3 license
 * http://www.lobsterdb.com/license
 */
Lobster.Composites.Composite.ComponentSet = function(args){
    this._stateArray = args.stateArray;
    this._classModel = args.classModel;
    this._composite = args.composite;
}

Lobster.Composites.Composite.ComponentSet.prototype = {

    item: function(key){
        for (var i = 0; i < this._stateArray.length; i++){
            if (this._stateArray[i][this._classModel.getKeyName()] == key){
                return new Lobster.Composites.Composite.Component({
                    state: this._stateArray[i]
                });
            }
        }    
    },
    
    getNew: function(){
        var classModel = this._classModel;
        var state = classModel.getTemplate();
        state[classModel.getKeyName()] = 
            this._composite.getComponentKey(classModel.getClassModelId);
        this._stateArray.push(state);
        return new Lobster.Composites.Composite.Component({
            state: state
        });
    },
    
    getAll: function(){
        var all = [];
        for (var i = 0; i < this._stateArray.length; i++){
            //all.push(
            //    new Lobster.Composites.Composite.
        }
    }

};